### DISCUSSION TITLE
Useful fixes for improve version will be included in debian 12

### DISCUSSION THREAD

**Discussion by Fantu** [memtest86plus/memtest86plus#253] [Q&A] [answered]
Hi, the debian 12 freeze is near and I was looking if there are important/useful fixes to apply (as debian patches), supposing that will be not new upstream version before the freeze.
I saw for example https://github.com/memtest86plus/memtest86plus/commit/485bfa46a3482a5c4190478e74202aa019bea30b that seems important fix (or more exactly workaround) to make memtest working on SuperMicro X10SDV.

Other fixes I suppose useful from a fast look to commits history, but I need to check and test, seems:
https://github.com/memtest86plus/memtest86plus/commit/03cd8d18985cd80757f709fa790f3cf1a746c1f5
https://github.com/memtest86plus/memtest86plus/commit/327495ec615e9ad6c16181484fe9157acd0b63d7

Any advice of other to apply, or to avoid to apply some of these is appreciated.

**Comment by x86fr** [comment]
Yes. Having the secureboot patch is also important. 

How many time do we have before the Debian 12 Freeze?

I can release v6.10 soon if needed to catch the Freeze.

**Comment by martinwhitaker** [comment]
The secure boot patch introduced a few bugs. I'll be submitting a PR to fix them soon.

**Comment by Fantu** [comment]
I don't have experience about secure boot part, I still don't looked/tested it and probably I don't have enough time recently (unless it's much easier and faster than I suppose)
I don't know if the other maintainer (Felix Zielcke) have more knowledge about it and time to try to add it before the freeze, I'll try to ask.
Soft freeze will be in few days (2023-02-12) and after only fixes should be done normally: https://release.debian.org/bookworm/freeze_policy.html

**Reply by x86fr** [reply]
@Fantu Is Memtest86+ considered high, medium or low priority packages? If low, it looks like it requires 10 days for migration so it will be too late for Feb, 12th.

**Reply by Fantu** [reply]
the priority is setted in changelog entry, normally medium (5 days for migration to testing) or low (10 days for migration), can complete the migration also in soft freeze, the only difference is that migration is fixed to 10 days.
anyway add features should be done before Feb, 12th, secure boot support is know to have bugs, I don't know how important and how difficult to solve (if fast and easy to solve can be solve also in freeze time). But what is probably missed a maintainer with knowledge and time to do the addition (and as long as there are no obstacles), but before exclude the possibility we need to wait Felix's answer first. Another possiblity is if others DD/DM want contribute and prepare a [merge request](https://salsa.debian.org/debian/memtest86plus/-/merge_requests).
But probably there is another more important thing, I don't know if there will be something blocking relating the signature procedure, I haven't looked at anything about it yet I don't know how it should be done and who should be contacted.

**Reply by x86fr** [reply]
PR has been submitted to solve last issues with SecureBoot patch.

It will be tested tomorrow then merged if everything's ok.

V6.10 could be released on Friday, 3rd.

Does this date sounds good to catch Debian 12 freeze?

**Reply by Fantu** [reply]
I don't want risk you will do a too fast release (with too short testing) to make possible enter in debian without even knowing if anyone will have enough time to do secure boot part in debian.
as mentioned, I don't have the knowledge at the moment, and therefore not even an idea of the time it would needed, if Felix (or other DD/DM if they want contributing) and have time, otherwise I don't know if I would have enough over the weekend (I also have other things to do in the queue) and if there will be any complications/expectations to be possible to sign the binary in the package

**Reply by x86fr** [reply]
OK. We'll do our best.

**Reply by debrouxl** [reply]
https://wiki.debian.org/IRC lists #debian-efi for UEFI and Secure Boot. Granted, IRC isn't necessarily the best medium for possibly nontrivial technical discussion, but that ought to be a good entry point to reach / get pointed to the relevant persons.
A lesser choice would be #debian-security , though "Secure" Boot is only a tiny part of (and a partial take on) computer security.

**Reply by Fantu** [reply]
thanks, I'll send a mail to the mailing list (https://lists.debian.org/debian-efi/) instead, I suppose is better and I'll try to write asking updated infos when I'll have time (this evening or next days), I'll put also other maintainer in cc.
I found these in wiki:
https://wiki.debian.org/SecureBoot
https://wiki.debian.org/SecureBoot/Discussion
but is better to know if are complete and updated before and I hadn't time to read them but only a very fast look few minutes previous evening, and it doesn't seem like a quick thing unfortunately (I could also be wrong). And it seems like DD is needed, Felix and I are just DM (or we'll probably need the procedure for a guest account).

**Comment by debrouxl** [comment]
It would be great to see a Secure Boot capable version of memtest86+ in Debian 12, but yeah, packager and maintainer time is a scarce resource.

**Comment by fezie** [comment]
@Fantu I would feel better, if there was an official upstream release with all important stuff applied and tested. Instead of doing it myself. Especially that short before the freeze.
Though it would be nice to have secure boot support in bookworm.

**Reply by Fantu** [reply]
@fezie hi, I saw now you did 6.10-1 upload.
about secure boot do you already have knowledge relating the creation of the certificate and the signature with it during the creation of the official packages on the debian servers?
about possible new debian revision upload adding secure boot support, have you time and want try to do it?

**Reply by x86fr** [reply]
IIRC on Debian, Memtest86+ will be set for a chainloading from grub, itself loaded by shim. This chain of trust requires the Memtest86+ .efi binary to be signed by Debian's CA (just like any Debian's signed Linux kernel). I don't know how this can be done but we're ready to help anyway.

On our side (that requires discussion and is unrelated to Debian), I'm not sure if the best option is to start a shim-review process with our CA to load a signed Memtest86+'s binary directly (to avoid potential security issue with grub like unimplemented CVE fixes), or to follow the same shim->grub->"kernel" process as in popular distros. Shim's doc states that they're not familiar with anything but GRUB2, but Memtest86+, being an "OS on its own", is a particular case. Adding GRUB in the chainloading process requires that we maintain a specific signed GRUB build only to load Memtest86+. That's perfectly fine for a distro like RH or Debian, but probably not for us. Maybe @vathpela can help at some point.

**Comment by x86fr** [comment]
We're doing extensive tests right now with #254 . The original secureboot patch has already been tested on numerous (> 300) motherboards without issues so far and #254 is polishing things, so I don't expect regressions. 

If all goes well, we'll do some more test on next days for a v6.10 release on Friday.

**Reply by Fantu** [reply]
there is a regression with build on i386: https://buildd.debian.org/status/fetch.php?pkg=memtest86%2B&arch=i386&ver=6.10-1&stamp=1675446455&raw=0

**Reply by martinwhitaker** [reply]
Why are you trying to build the 64-bit image for i386?

**Reply by martinwhitaker** [reply]
I don't have a 32-bit system to test on, but this should fix it
````
--- a/build64/Makefile
+++ b/build64/Makefile
@@ -78,11 +78,11 @@ boot/header.o : | ../boot/sbat.csv
 
 boot/startup.o: ../boot/startup64.S ../boot/boot.h
        @mkdir -p boot
-       $(CC) -x assembler-with-cpp -c -I../boot -o $@ $<
+       $(CC) -m64 -x assembler-with-cpp -c -I../boot -o $@ $<
 
 boot/%.o: ../boot/%.S ../boot/boot.h app/build_version.h
        @mkdir -p boot
-       $(CC) -x assembler-with-cpp -c -I../boot -Iapp -o $@ $<
+       $(CC) -m64 -x assembler-with-cpp -c -I../boot -Iapp -o $@ $<
 
 boot/efisetup.o: ../boot/efisetup.c
        @mkdir -p boot
````

**Reply by Fantu** [reply]
I keep that from pcmemtest packaging did by Felix and taken as start for v6 packaging, I think was ok keep, for example as memtest is not installed in debian but "out" of it and there can be 32bit debian installation on computer 64 bit capable and can also have more ram installed that limit supported by 32 bit debian installation. Or you think is not a good idea crossarch build and/or keep both 32 and 64 bit binary of memtest?

**Reply by martinwhitaker** [reply]
I guess I'm just used to the way we do things in Mageia - an arch'd package only contains binaries for that arch. But no reason not to do cross arch builds if you want to.

**Reply by x86fr** [reply]
Can you handle this on Debian's build process for v6.10 (and we will push that flag on Makefile later)? I can still push this, change the actual v6.10 tag and release, but it's not extremely clean.

**Reply by Fantu** [reply]
replace the tag is not good (for example if other distro already released 6.10), on debian we can add the needed fixes in debian/patches until the next upstream version

**Reply by fezie** [reply]
> Why are you trying to build the 64-bit image for i386?

A 64 bit System can run a 32bit System. But a 64bit EFI can't run 32bit EFI binaries.
That's why grub also has 32bit and 64bit EFI files on both amd64 and i386 Debian architectures.
We could maybe make memtest86+ multiarch compatible. But how we do it now is the old way.

**Comment by x86fr** [comment, marked as answer]
v6.10 released! 🎉🎊
